%\addcontentsline{toc}{chapter}{Development Process}
\section{Provided data}
The data for this project was produced by a contracted researcher who had moved onto another project. Because of this there was limited opportunity to talk with them about the data, how it was produced and what needed doing to it. Unfortunately, the data did not include any form of documentation, or annotation as to how it was produced, the only metadata available was the filenames. 

This lead to a lot of confusion as it wasn't immediately obvious what each bit of data was and what it meant. Examining the data and learning about the biology behind it took several weeks, and ascertaining what bits of data were actually relevant to the project took even longer. Eventually it was determined that for each species there were three files that would need to be incorporated into the project.

\begin{itemize}
  \item Raw assembled contigs of DNA, these would be used to get a genes position in the genome, and it's surrounding bases. 
  \item Protein sequences found from using BLAST, annotated with Gene Ontology ID's.
  \item Coding sequences, these are the nucleotide sequences that coded for the proteins. 
\end{itemize}

Linking this data to the Candida Genome Database (CGD) was a key bit of information that was missing. To do this the annotated proteins\cite{albicans} for \textit{C. albicans} were downloaded, then turned into a reference database with Diamond. With this the three species could then have their coding sequences aligned against \textit{C. albicans} to produce accession ID's from the Candida Genome Database. The script that performed this step is included in the project as \texttt{run\_diamond.sh}.

Several sets of data were then required to map these proteins to the Candida Genome Database, which would allow the researchers to easily compare the genes to well annotated species. First was a file of GeneOntology\cite{geneontology} annotations for all the proteins stored in the Candida Genome Database\cite{cgd-proteins}. This file conveniently also stores all the accession ID's for those proteins. This is what was used to map the alignment results to the Candida Genome Database, by selecting the accession ID from the Diamond alignment results, and looking that up in this data set, the Candida Genome ID was able to be found.

In addition to this was a mapping file that mapped Candida Genome Database ID's to UniProt\cite{uniprot} ID's. With a small JavaScript script and some manual cleaning with a text editor, a complete mappings file was able to be produced in JSON, ready to be read in by the importation script. This allows for genes found to be linked to other well known genes in the Candida Genome Database and UniProt.
